
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/dataplane/standalone/proto";

enum L2mcGroupAttr {
  L2MC_GROUP_ATTR_UNSPECIFIED = 0;
  L2MC_GROUP_ATTR_L2MC_OUTPUT_COUNT = 1;
  L2MC_GROUP_ATTR_L2MC_MEMBER_LIST = 2;
}

enum L2mcGroupMemberAttr {
  L2MC_GROUP_MEMBER_ATTR_UNSPECIFIED = 0;
  L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID = 1;
  L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID = 2;
  L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP = 3;
}

message CreateL2mcGroupRequest {
  option (sai_type) = OBJECT_TYPE_L2MC_GROUP;
  uint64 switch = 1;
}

message CreateL2mcGroupResponse {
  uint64 oid = 1;
}

message RemoveL2mcGroupRequest {
  uint64 oid = 1;
}

message RemoveL2mcGroupResponse {}

message GetL2mcGroupAttributeRequest {
  uint64 oid = 1;
  repeated L2mcGroupAttr attr_type = 2;
}

message GetL2mcGroupAttributeResponse {
  L2mcGroupAttribute attr = 1;
}

message CreateL2mcGroupMemberRequest {
  option (sai_type) = OBJECT_TYPE_L2MC_GROUP_MEMBER;
  uint64 switch = 1;
  optional uint64 l2mc_group_id = 2 [(attr_enum_value) = 1];
  optional uint64 l2mc_output_id = 3 [(attr_enum_value) = 2];
  optional bytes l2mc_endpoint_ip = 4 [(attr_enum_value) = 3];
}

message CreateL2mcGroupMemberResponse {
  uint64 oid = 1;
}

message RemoveL2mcGroupMemberRequest {
  uint64 oid = 1;
}

message RemoveL2mcGroupMemberResponse {}

message GetL2mcGroupMemberAttributeRequest {
  uint64 oid = 1;
  repeated L2mcGroupMemberAttr attr_type = 2;
}

message GetL2mcGroupMemberAttributeResponse {
  L2mcGroupMemberAttribute attr = 1;
}

service L2mcGroup {
  rpc CreateL2mcGroup(CreateL2mcGroupRequest)
      returns (CreateL2mcGroupResponse) {}
  rpc RemoveL2mcGroup(RemoveL2mcGroupRequest)
      returns (RemoveL2mcGroupResponse) {}
  rpc GetL2mcGroupAttribute(GetL2mcGroupAttributeRequest)
      returns (GetL2mcGroupAttributeResponse) {}
  rpc CreateL2mcGroupMember(CreateL2mcGroupMemberRequest)
      returns (CreateL2mcGroupMemberResponse) {}
  rpc RemoveL2mcGroupMember(RemoveL2mcGroupMemberRequest)
      returns (RemoveL2mcGroupMemberResponse) {}
  rpc GetL2mcGroupMemberAttribute(GetL2mcGroupMemberAttributeRequest)
      returns (GetL2mcGroupMemberAttributeResponse) {}
}
